### ISSUE TITLE
Add ethereumjs-testrpc integration

### ISSUE THREAD

**Issue by flyswatter** [trufflesuite/truffle#67]
For `development` and `test` environments, it's unlikely anyone wants to run their code on the real Ethereum blockchain, so they're either going to set up their own private Ether blockchain, or they're going to use a tool like [EtherSim](https://github.com/tcoulter/EtherSim).

Since EtherSim can be passed as a provider to `web3.setProvider()`, it seems like it would be easy to swap in instead of `new Web3.providers.HttpProvider(env_defined_address)` based on an environment or command.

This could either be defined in the `app.json` file, or maybe (as @tcoulter suggested) as an alternate command to `serve`, like "`develop`".

Let's talk about how we might expose this functionality, so that it's easily available for all new Ether apps!

**Comment by tcoulter** [comment]
Technically speaking, EtherSim *itself* as a provider won't work as there's no way to connect the provider in node to the provider that'd run in the browser (aside: I think that feature of EtherSim is lofty and not thought through). That said, having one command that runs EtherSim in the background and takes care of the provider configuration is a really nice feature.

**Comment by flyswatter** [comment]
Yeah, just running it in the background (assuming it's installed, or even including it as a dev dependency of truffle) should be pretty simple.

On that side note, wary of getting too far off track, why is EtherSim as a provider not well thought out? It seems like that would be fairly simple, and should be isomorphic, so shouldn't *require* the node environment.  But I guess that's an issue for EtherSim.

**Comment by flyswatter** [comment]
So let's say we add a `truffle develop` keyword. Would it automatically pass the current `rpc.port` value to `EtherSim` as its port, or would we add a new `EtherSim` config hash to the `app.json`?

**Comment by tcoulter** [comment]
> why is EtherSim as a provider not well thought out?

Your dapp and Truffle run in two different execution contexts: The first in the browser, the second in Node. Since EtherSim is basically a blockchain in memory, there's no way to share an instance of EtherSim between the two contexts. Instead, you would just load up EtherSim as its own entity and have both Truffle and your dapp communicate with EtherSim over HTTP, just like any other HttpProvider. Note that because EtherSim is a blockchain-in-memory, we *need* to share one instance as deploying to one instance of EtherSim won't share state with another. 

Answering the question more generally, this means that EtherSim as a provider (or, "as a library") is only useful within the execution context that instantiates it. And in that case it's effectively a nice-to-have that makes a few things easier, like not having to deal with port numbers, but otherwise doesn't provide a whole lot of added value. For instance:

As a web server (pseudo-code):
```
var Server = require("ethersim/server");
Server.startServer(8545);
web3.setProvider(new web3.providers.HttpProvider("http://localhost:8545"));
```

vs., as a library (pseudo-code):
```
var provider = require("ethersim/provider");
web3.setProvider(provider);
```

I can imagine some cases where it's important to not have a port number, but I think those cases are very few. Also, Iuri's implementation of it here https://github.com/iurimatias/EtherSim/commit/0d17f6d5bbcb032272417a41d8e688f28023474d (i.e., directly accessing methods through the Manager object, and wanting to support not needing to provide a callback) is bad code and can be done in much better ways without opening up the possibility for new error cases (for instance, this line is fundamentally broken and requires a callback: https://github.com/iurimatias/EtherSim/commit/0d17f6d5bbcb032272417a41d8e688f28023474d#diff-b840698050e033004db3ac04392b9ea1L81).

So, in short, I can see the value, but there's not much value. And in my opinion the values not worth the mess that's happening in Iuri's code.

**Comment by tcoulter** [comment]
> So let's say we add a truffle develop keyword. Would it automatically pass the current rpc.port value to EtherSim as its port, or would we add a new EtherSim config hash to the app.json

Well, we'd make it do that, ya. But to do so, truffle would have to do things in this order:

1. Start Ethersim on some port
1. Deploy contracts to EtherSim (optionally recompile compile at this step, or compile if they're not yet)
1. Change the `config` value to point to the Ethersim we loaded up
1. Rebuild the app with that `config` (will happen automatically if we run `truffle serve`) 
1. Viola! The Truffle bootstrapping that happens within your dapp will take care of the rest (i.e., pointing at the host and port of the EtherSim we started in step (1), since it was in the config)

**Comment by flyswatter** [comment]
That sounds good! I'll take this opportunity to graze over the Truffle internals to familiarize myself with where these things happen.

- Dan

**Comment by tcoulter** [comment]
Hmm, I'm not sure I agree with my statements above. I didn't do a full reread, but looks like I was plain wrong. That said, I'm now thinking this:

-------------------

I’m thinking of integrating the testrpc into Truffle. So you could do something like this:

```
$ truffle migrate —dry-run http://localhost:8545
```

Where you point it to the main chain and it migrates off a fork. Similarly, you wouldn't have to set up your own testrpc for `truffle test`.

**Comment by jvah** [comment]
Just in case this is helpful, I'm using my own package.json and a script like this to run tests:

```
    "test": "testrpc & truffle test ; RETVAL=$? ; JOBS=`jobs -p` ; if [ -n \"$JOBS\" ]; then kill $JOBS; fi ; exit $RETVAL"
```

It attempts to start testrpc (which will fail in case there already is a client listening to the same port), then run the tests and finally kill testrpc in case it is still running, before returning the mocha return code.

**Comment by tcoulter** [comment]
This ticket is old. We're tracking this in #490. Still a good idea now more than ever though.
